main/src/ca/uwaterloo/flix/language/phase/Deriver.scala

```diff
@@ -336,6 +338,214 @@ object Deriver extends Phase[KindedAst.Root, KindedAst.Root] {
       KindedAst.MatchRule(pat, guard, exp)
   }
 
+  /**
+    * Creates an Order instance for the given enum.
+    *
+    * {{{
+    * enum E[a] with Order {
+    *   case C1
+    *   case C2(a)
+    *   case C3(a, a)
+    * }
+    * }}}
+    *
+    * yields
+    *
+    * {{{
+    * instance Order[E[a]] with Order[a] {
+    *   pub def compare(x: E[a], y: E[a]): Comparison = {
+    *     let indexOf = e -> match e {
+    *       case C0(_) -> 0
+    *       case C1(_) -> 1
+    *       case C2(_) -> 2
+    *     };
+    *     match (x, y) {
+    *       case (C0, C0) => Comparison.EqualTo
+    *       case (C1(x0), C1(y0)) => Order.compare(x0, y0)
+    *       case (C2(x0, x1), C2(y0, y1)) => Order.compare(x0, y0) `Order.thenCompare` lazy Order.compare(x1, y1)
+    *       case _ => Order.compare(indexOf(x), indexOf(y))
```

Comment:
I guess we can leave a ticket to do this more efficiently.

Reply:
Improving the performance of compare will dramatically affect the performance of sets and maps.

Reply:
But this probably requires a multi-pronged approach. Starting with something simple that works is probably a good idea.

Reply:
Yes. It also gives us a baseline for comparing future improvements.